tests/basic.rs

```diff
@@ -1082,3 +1082,13 @@ fn no_sender_returns_disconnected() {
     drop(addr);
     assert!(!ctx.weak_address().is_connected());
 }
+
+#[tokio::test]
+async fn receive_future_can_dispatch_in_one_poll() {
+    let (addr, ctx) = Context::<Greeter>::new(None);
+
+    let _ = addr.send(Hello("world")).split_receiver().await;
+    let receive_future = ctx.next_message();
+
+    assert!(receive_future.now_or_never().is_some())
```

Comment:
I think the race here would be:
- now or never returns None
- message is sent 
- drop is called
- message is found and sent, being reordered

Reply:
Therefore, this is better in most but not all cases than just dropping

Reply:
Sorry, I don't fully understand.

I think this test proves that, once you have a `ReceiveFuture`, it can complete in a single poll, even if it has never been polled yet. Isn't this enough?

I guess we are not testing the case in which it is `Waiting`.

Reply:
What do you think of other test I just added?

Reply:
>  think this test proves that, once you have a ReceiveFuture, it can complete in a single poll, even if it has never been polled yet. Isn't this enough?

I agree, but this is not the race which I was describing. I might be able to write a test tomorrow showing exactly what I mean, though not with now_or_never, but with manual polling emulating it. 

I think the other test is also valuable, but doesn't quite test what I meant.

Reply:
I think I understand what you are trying to showcase but I am not sure it is necessary.

As it stands, we know that messages can be re-ordered if the `ReceiveFuture` is dropped. That is also documented on the type and users can decide to accept this reordering and simply drop a `ReceiveFuture`.

If they really care about ordering, they can do what we do in `Context::select` and call `now_or_never` on the future.

The tests we add here prove that, no matter what state the `ReceiveFuture` is in (newly constructed or already polled), it will always yield the message IF it has one stored internally. With this test passing, we should be able to rely on `now_or_never` to basically be a 1-to-1 replacement for `cancel`.

Reply:
I think it comes down to what kind of ordering we would like to guarantee when multiple actors are involved. `now_or_never` has equivalent ordering with `Drop` *and even* `cancel` if there is only one receiving actor.
 
 ```rust
 let (addr, mut ctx) = Context::new(None);

let mut a = Cow { name: "Alfred" };

let mut rx_a = ctx.next_message();
let mut fut_ctx = std::task::Context::from_waker(noop_waker_ref());

// Pretend this is the poll in now_or_never...
assert!(rx_a.poll_unpin(&mut fut_ctx).is_pending());

// This will fulfill the waiting rx in rx_a
assert!(addr.send(Hello("first message")).split_receiver().now_or_never().is_some());

// This will go to the queue
assert!(addr.send(Hello("second message")).split_receiver().now_or_never().is_some());

// Pretend this receive is from another actor...
ctx.tick(ctx.next_message().await, &mut a).await;

// Pretend this is the drop in now_or_never
drop(rx_a);

// This receive is also from the other actor...
ctx.tick(ctx.next_message().await, &mut a).await;
```

In this case, the second actor has now handled the message out of order.

I don't actually think this is much of a problem, on second thought, but we should mention how ordering interacts with multiple actors, just in case, although I do not expect multiple actors on the same address to be relying on external message ordering.

Reply:
I see. What is missing for me in the above case is to do something with `rx_a` after we have initially polled it. I think it is okay for things to behave weirdly if you have half-polled futures around that are not interacted with.

1. If you `Drop` the `rx_a` future before the first send, it doesn't have a message and thus no re-ordering takes place
2. If you `Drop` it afterwards, the message gets re-inserted and re-ordering might take place
3. If you poll it again after the first send (with `now_or_never`), the message will be delivered to you and you can pass it to an actor.

(3) is what we are doing in the `select` function now which I think is functionally equivalent to what we had with `cancel`.



> I don't actually think this is much of a problem, on second thought, but we should mention how ordering interacts with multiple actors, just in case, although I do not expect multiple actors on the same address to be relying on external message ordering.

I don't think we want to make any guarantees in regards to message ordering with our work-stealing scheduling. That would be kind of bizarre.

Reply:
> I think it is okay for things to behave weirdly if you have half-polled futures around that are not interacted with.

Fair enough. There was a similar discussion in flume at some point, if I remember correctly, and we came to a similar decision.

> If you poll it again after the first send (with now_or_never), the message will be delivered to you and you can pass it to an actor.

Technically, the message will be delivered to you whether or not you poll it more than once. As soon as you poll it once it will be delivered to you if the sender is given an opportunity.

> I don't think we want to make any guarantees in regards to message ordering with our work-stealing scheduling. That would be kind of bizarre.

Agreed.

Reply:
Okay. I think this is ready to merge then, right?

Reply:
I think so, yes. Maybe as part of #121 we can mention the ordering